nano/node/distributed_work.cpp

```diff
@@ -0,0 +1,367 @@
+						if (!ec)
+						{
+							boost::beast::http::async_read (connection->socket, connection->buffer, connection->response, [this_l, connection](boost::system::error_code const & ec, size_t bytes_transferred) {
+								if (!ec)
+								{
+									if (connection->response.result () == boost::beast::http::status::ok)
+									{
+										this_l->success (connection->response.body (), connection->address);
+									}
+									else
+									{
+										this_l->node.logger.try_log (boost::str (boost::format ("Work peer responded with an error %1% %2%: %3%") % connection->address % connection->port % connection->response.result ()));
+										this_l->failure (connection->address);
+									}
+								}
+								else if (ec == boost::system::errc::operation_canceled)
+								{
+									// The only case where we send a cancel is if we preempt stopped waiting for the response
+									this_l->cancel (connection);
+									this_l->failure (connection->address);
+								}
+								else
+								{
+									this_l->node.logger.try_log (boost::str (boost::format ("Unable to read from work_peer %1% %2%: %3% (%4%)") % connection->address % connection->port % ec.message () % ec.value ()));
+									this_l->failure (connection->address);
+								}
+							});
+						}
+						else
+						{
+							this_l->node.logger.try_log (boost::str (boost::format ("Unable to write to work_peer %1% %2%: %3% (%4%)") % connection->address % connection->port % ec.message () % ec.value ()));
+							this_l->failure (connection->address);
+						}
+					});
+				}
+				else
+				{
+					this_l->node.logger.try_log (boost::str (boost::format ("Unable to connect to work_peer %1% %2%: %3% (%4%)") % connection->address % connection->port % ec.message () % ec.value ()));
+					this_l->failure (connection->address);
+				}
+			});
+		}
+	}
+}
+
+void nano::distributed_work::cancel (std::shared_ptr<nano::work_peer_request> connection)
+{
+	auto this_l (shared_from_this ());
+	auto cancelling_l (std::make_shared<nano::work_peer_request> (node.io_ctx, connection->address, connection->port));
+	cancelling_l->socket.async_connect (nano::tcp_endpoint (cancelling_l->address, cancelling_l->port), [this_l, cancelling_l](boost::system::error_code const & ec) {
+		if (!ec)
+		{
+			std::string request_string;
+			{
+				boost::property_tree::ptree request;
+				request.put ("action", "work_cancel");
+				request.put ("hash", this_l->root.to_string ());
+				std::stringstream ostream;
+				boost::property_tree::write_json (ostream, request);
+				request_string = ostream.str ();
+			}
+			auto request (std::make_shared<boost::beast::http::request<boost::beast::http::string_body>> ());
+			request->method (boost::beast::http::verb::post);
+			request->set (boost::beast::http::field::content_type, "application/json");
+			request->target ("/");
+			request->version (11);
+			request->body () = request_string;
+			request->prepare_payload ();
```

Comment:
These seven lines are duplicated in distributed_work::start_work (). Maybe use this refactoring as an opportunity to extract this into a `prepare_request` function or something similar.

Reply:
Definitely, fits well into the `work_peer_request` class imo, despite not needing anything from it.